src/lib/getElementType.js

```diff
@@ -0,0 +1,41 @@
+import _ from 'lodash'
+
+/**
+ * Returns a createElement() type based on the props of the Component.
+ * Useful for calculating what type a component should render as.
+ *
+ * @param {function} Component A function or ReactClass.
+ * @param {object} props A ReactElement props object
+ * @param {object} [typeMap] Maps props (keys) to element types (values).
+ * @returns {string|function} A ReactElement type
+ */
+function getElementType(Component, props, typeMap) {
+  const { defaultProps = {} } = Component
+
+  // ----------------------------------------
+  // user defined element type
+
+  if (props.as && props.as !== defaultProps.as) return props.as
```

Comment:
OK, this is still required.  The element type is chosen from this priority:

1. User defined an `as` prop
1. getElementType() typeMap arg mapped a prop to an element type:  `{ onClick: 'a' }`
1. Inferred type: `href => a'
1. defaultProp.as
1. `div`

When there is a defaultProp, the `props.as` will be set.  However, this _is not_ the user's `as` value.  It is the component default.  If we remove the `props.as !== defaultProps.as` comparison, then the defaultProp.as is always used as the first priority since it is copied to `props.as`.

So, we only want to use the `props.as` if the user defined one (i.e. it does not equal the default).

**Image.js**

- default - `img`
- maps - `wrapped => div`
- infers - `href => a`

If the props.as is used, it will always be an `img`.  Even if the user uses `wrapped`, which should be a div or `href` which should wrap the image in an `a` tag.

```jsx
render() {
  const ElementType = getElementType(Image, props, {
    wrapped: 'div',
  })

  if (ElementType === 'img') {
    return <ElementType {...rootProps} {...imgTagProps} />
  }

  return (
    <ElementType {...rootProps} href={href}>
      <img {...imgTagProps} />
    </ElementType>
  )
}

Image.defaultProps = {
  as: 'img',
}
```